### PR TITLE
Fix float16 L2 normalize gradients via float32 accumulation

### LINKED ISSUE #23546: Bug: keras.src.ops.normalize() produces -inf gradients for small non-zero float16 inputs for the torch backend
### Description
`keras.src.ops.normalize()`'s L2 path performs its squared-norm computation, epsilon clamp, and `rsqrt` entirely in the input's original dtype. For `float16` inputs, this causes the gradient to become non-finite (`inf`/`NaN`) even for inputs that are not close to zero, because `epsilon**2` underflows to `0.0` in `float16` and `rsqrt`'s derivative overflows `float16`'s range at magnitudes well within normal usage. The example below shows a gradient of `-inf` for an input array filled with `0.01`, versus a small finite gradient in `float32`.

### Reproduction 
```
import os
os.environ["KERAS_BACKEND"] = "torch"

import numpy as np
import torch

from keras.src.ops import normalize

x_np = np.full((4, 2), 0.01, dtype=np.float16) 

x = torch.tensor(x_np, requires_grad=True)
y = normalize(x, axis=-1, order=2)
y.sum().backward()
grad = x.grad.detach().cpu().numpy()

print(f"gradient:\n{grad}")
# gradient:
# [[-inf -inf]
#  [-inf -inf]
#  [-inf -inf]
#  [-inf -inf]]

```

#### Output with float32
```
gradient:
[[7.6293945e-06 7.6293945e-06]
 [7.6293945e-06 7.6293945e-06]
 [7.6293945e-06 7.6293945e-06]
 [7.6293945e-06 7.6293945e-06]]
```

#### Expected Behavior
I would expect the gradient not to be -inf because it is far from that. It would be more logical for the gradient to be 0 everywhere. 

### Possible Fix
A possible fix would be to upcast the input to at least float32 before computing the squared norm, epsilon clamp, and rsqrt, and only casting back to the original dtype at the end. The following shows an example patch

```
diff --git a/keras/src/ops/nn.py b/keras/src/ops/nn.py
index c4d1b396a..ad6d7fab0 100644
--- a/keras/src/ops/nn.py
+++ b/keras/src/ops/nn.py
@@ -2722,14 +2722,23 @@ def _normalize(x, axis=-1, order=2, epsilon=None):
     if 2 == order:
         # A special case: L2 normalization with `x * rsqrt(...)`
         # instead of `x / sqrt(...)`. Clamp the squared norm before the
         # rsqrt so zero vectors get a finite gradient.
+        original_dtype = getattr(backend, 'standardize_dtype', lambda d: d)(x.dtype)
+        compute_dtype = getattr(backend, 'result_type', lambda a, b: 'float32')(x.dtype, 'float32')
+        x_cast = backend.cast(x, dtype=compute_dtype)
         square_sum = backend.numpy.sum(
-            backend.numpy.square(x), axis=axis, keepdims=True
+            backend.numpy.square(x_cast), axis=axis, keepdims=True
         )
+        # compute epsilon squared in compute dtype to avoid underflow
+        eps_sq = backend.cast(epsilon * epsilon, dtype=compute_dtype)
         inv_norm = backend.math.rsqrt(
-            backend.numpy.maximum(square_sum, epsilon * epsilon)
+            backend.numpy.maximum(square_sum, eps_sq)
         )
-        return x * inv_norm
+        out = x_cast * inv_norm
+        return backend.cast(out, dtype=original_dtype)
+
     norm = backend.linalg.norm(x, ord=order, axis=axis, keepdims=True)
     denom = backend.numpy.maximum(norm, epsilon)
     return backend.numpy.divide(x, denom)
```

### AI Disclosure
I used an LLM based approach for research to find this bug and I also used an agentic approach to find a plausible fix. I reviewed the found bug by checking if the behavior is still active on the most current version and agreeing that this behavior is indeed unintended. I also reviewed the proposed patch and take full responsibility for it.

### Environment
- Keras version: 78d44013f at the master branch
- Backend: NumPy (`KERAS_BACKEND=numpy`)
- Python version: Python 3.11.2
- OS: Debian GNU/Linux 12 (bookworm)

### PR DESCRIPTION
**Pull request by indrajeetapache**
## Summary

Fixes #23546.

`ops.normalize(..., order=2)` computed `square` / `epsilon**2` / `rsqrt` in the input dtype. On `float16`, the `rsqrt` derivative overflows for ordinary magnitudes (e.g. arrays filled with `0.01`), so Torch (and other autodiff backends) return `-inf` gradients. The same values stay finite in `float32`.

#23093 already clamped before `rsqrt` to fix zero-vector NaNs, but the half-precision upcast never landed. This PR applies the same pattern Keras already uses in `_rms_normalization` / layer norm:

- `compute_dtype = result_type(x.dtype, "float32")`
- run the L2 path in `compute_dtype`
- cast the output back to the input dtype

Also adds `test_normalize_l2_float16_gradients`, which checks that float16 grads are finite and match the float32 reference.

## Test plan

- [x] Existing `test_normalize_l2_zero_vector_gradients` still covers the #23075 case
- [x] New float16 gradient test for tensorflow / jax / torch
- [x] CI: `keras/src/ops/nn_test.py` on torch / tensorflow / jax

## Contributor Agreement

**Please review our [PR Contribution Policy](https://github.com/keras-team/keras/issues/23601) and [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] This PR is linked to an issue that has been assigned to me (see `Fixes #xxx` above).
- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed. PRs without a linked, assigned issue will be converted to draft.

### CODE REVIEW

**Comment by JyotinderSingh** [comment]
Please agree to the contributors agreement to allow the CI to run.

**Comment by indrajeetapache** [comment]
@JyotinderSingh Thank you for ping, it's now done.

File: keras/src/ops/nn.py [hunk 1]
```diff
@@ -2735,14 +2735,19 @@ def _normalize(x, axis=-1, order=2, epsilon=None):
...
+        compute_dtype = backend.result_type(x.dtype, "float32")
+        x = backend.cast(x, compute_dtype)
         square_sum = backend.numpy.sum(
             backend.numpy.square(x), axis=axis, keepdims=True
         )
         inv_norm = backend.math.rsqrt(
             backend.numpy.maximum(square_sum, epsilon * epsilon)
         )
-        return x * inv_norm
+        return backend.cast(x * inv_norm, original_dtype)
```
Comment (JyotinderSingh) [inline keras/src/ops/nn.py]: The cast to `original_dtype` truncates the float result for integer and boolean inputs. Before this PR, the torch, jax and numpy backends return float32 here, and the tensorflow backend raises an error for an integer input. Now each backend returns zeros:

```python
keras.ops.normalize(keras.ops.convert_to_tensor([[1, 2, 3], [4, 5, 6]]))
# before (torch/jax): float32 [[0.267, 0.535, 0.802], [0.456, 0.570, 0.684]]
# now:                int32   [[0, 0, 0], [0, 0, 0]]
```

This example is in the docstring of `normalize`. The change also breaks the docstring example of `UnitNormalization`. `keras.losses.cosine_similarity` and `Dot(normalize=True)` return zeros for integer inputs. A boolean input comes back without a change. No test uses an input that is not a float, and the test suite does not find this error. Only a float input keeps its dtype. The suggestion below makes this change.

File: keras/src/ops/nn.py [hunk 2]
```diff
@@ -2735,14 +2735,19 @@ def _normalize(x, axis=-1, order=2, epsilon=None):
     if 2 == order:
         # A special case: L2 normalization with `x * rsqrt(...)`
         # instead of `x / sqrt(...)`. Clamp the squared norm before the
-        # rsqrt so zero vectors get a finite gradient.
+        # rsqrt so zero vectors get a finite gradient. Compute in at least
+        # float32 so half-precision inputs do not underflow `epsilon**2` or
+        # overflow the rsqrt derivative (see #23546).
+        original_dtype = backend.standardize_dtype(x.dtype)
+        compute_dtype = backend.result_type(x.dtype, "float32")
```
Comment (JyotinderSingh) [inline keras/src/ops/nn.py]: `backend.result_type` gives float32 for a float64 input on each backend except tensorflow. On the torch backend, this line changes a float64 input to float32 for the full calculation. The function also raises an error for a float8 input, and a float8 input works on the jax backend before this PR.

Use float32 only for a float16 or a bfloat16 output dtype. `_lgamma` in `keras/src/ops/math.py` uses this method. The suggestion below does the same.

File: keras/src/ops/nn.py [hunk 3]
```diff
@@ -2735,14 +2735,19 @@ def _normalize(x, axis=-1, order=2, epsilon=None):
...
             backend.numpy.square(x), axis=axis, keepdims=True
         )
         inv_norm = backend.math.rsqrt(
             backend.numpy.maximum(square_sum, epsilon * epsilon)
         )
-        return x * inv_norm
+        return backend.cast(x * inv_norm, original_dtype)
     norm = backend.linalg.norm(x, ord=order, axis=axis, keepdims=True)
     denom = backend.numpy.maximum(norm, epsilon)
     return backend.numpy.divide(x, denom)
```
Comment (JyotinderSingh) [inline keras/src/ops/nn.py]: The `order != 2` branch has the same problem in half precision. With this PR, a float16 input of `0.01` still gives an `-inf` gradient for `order=3`. The result is the same on the torch, jax and tensorflow backends. A bfloat16 input gives a finite but incorrect gradient. For `order=3` and a magnitude of `1e-4` on the jax backend, the gradient is `384`. The float32 gradient is `0.0068`.

Move the dtype code above the branch. This change corrects each order and the two comments above:

```suggestion
    original_dtype = backend.standardize_dtype(x.dtype)
    if backend.is_float_dtype(original_dtype):
        output_dtype = original_dtype
    else:
        output_dtype = backend.result_type(original_dtype, backend.floatx())
    # Compute half-precision inputs in float32 so the norm and its gradient
    # do not lose range or precision (see #23546).
    compute_dtype = (
        "float32" if output_dtype in ("float16", "bfloat16") else output_dtype
    )
    x = backend.cast(x, compute_dtype)
    if 2 == order:
        # A special case: L2 normalization with `x * rsqrt(...)`
        # instead of `x / sqrt(...)`. Clamp the squared norm before the
        # rsqrt so zero vectors get a finite gradient.
        square_sum = backend.numpy.sum(
            backend.numpy.square(x), axis=axis, keepdims=True
        )
        inv_norm = backend.math.rsqrt(
            backend.numpy.maximum(square_sum, epsilon * epsilon)
        )
        outputs = x * inv_norm
    else:
        norm = backend.linalg.norm(x, ord=order, axis=axis, keepdims=True)
        denom = backend.numpy.maximum(norm, epsilon)
        outputs = backend.numpy.divide(x, denom)
    return backend.cast(outputs, output_dtype)
```

`Normalize.compute_output_spec` needs the same output dtype. The review body has that change.

File: keras/src/ops/nn_test.py [hunk 4]
```diff
@@ -2733,6 +2733,55 @@ def f(x):
...
+            y32 = knn.normalize(x32, axis=-1, order=2)
+            y32.sum().backward()
+            expected_grad = x32.grad
+        else:
+            self.skipTest("Gradient test requires tensorflow, jax or torch.")
+
+        x_grad = ops.convert_to_numpy(x_grad)
+        expected_grad = ops.convert_to_numpy(expected_grad)
+        self.assertTrue(np.isfinite(x_grad).all())
+        self.assertAllClose(x_grad.astype("float32"), expected_grad, atol=1e-3)
```
Comment (JyotinderSingh) [inline keras/src/ops/nn_test.py]: Each value in the input is equal. For a row with equal values, `sum(normalize(x))` is at its maximum, so the correct gradient is exactly zero. `expected_grad` contains only zeros. The `assertAllClose` call passes for each gradient near zero, and a version with `ops.stop_gradient` also passes. On the tensorflow backend, the code before this PR gives finite gradients for this input. The check for finite values does not test the error on that backend.

This version uses different values in each row and a weight for each value. The weights make the loss depend on the direction of each row, so the correct gradient is not zero for any order. The test also covers `order=1`, `order=3` and bfloat16. It fails with the code before this PR and with the current PR, and it passes with the suggestion for `nn.py`:

```suggestion
    @parameterized.product(order=[1, 2, 3], dtype=["float16", "bfloat16"])
    def test_normalize_half_precision_gradients(self, order, dtype):
        # Half-precision gradients must match float32 for ordinary magnitudes
        # (see #23546). The weights make the loss depend on the direction of
        # each row. Without them, `sum(normalize(x))` is constant for `order=1`
        # with positive values and at its maximum for a row with equal values,
        # so its gradient is exactly zero and the comparison would be vacuous.
        if backend.backend() not in ("tensorflow", "jax", "torch"):
            self.skipTest("Gradient test requires tensorflow, jax or torch.")
        x_np = np.array([[0.001, 0.002], [0.003, 0.0005]], dtype="float32")
        weights = np.array([[1.0, -2.0], [0.5, -3.0]], dtype="float32")

        def loss_fn(x):
            y = knn.normalize(x, axis=-1, order=order)
            return ops.sum(ops.multiply(y, weights))

        def grad(x_dtype):
            if backend.backend() == "tensorflow":
                import tensorflow as tf

                x = tf.Variable(tf.cast(x_np, x_dtype))
                with tf.GradientTape() as tape:
                    loss = loss_fn(x)
                return tape.gradient(loss, x)
            if backend.backend() == "jax":
                import jax

                return jax.grad(loss_fn)(jax.numpy.array(x_np, dtype=x_dtype))
            import torch

            x = torch.tensor(
                x_np, dtype=getattr(torch, x_dtype), requires_grad=True
            )
            loss_fn(x).backward()
            return x.grad

        x_grad = ops.convert_to_numpy(grad(dtype)).astype("float32")
        expected_grad = ops.convert_to_numpy(grad("float32"))
        self.assertTrue(np.isfinite(x_grad).all())
        self.assertAllClose(x_grad, expected_grad, atol=1e-2, rtol=1e-2)
```

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Thanks for the fix. The float32 upcast corrects the case in #23546. Four items need a change before this PR goes in.

1. **The cast to the input dtype breaks integer and boolean inputs.** See the comment on line 2750.
2. **The `order != 2` branch has the same error in half precision.** See the comment on lines 2735 to 2753. The suggestion there corrects both branches.
3. **`Normalize.compute_output_spec` sets float32 for each input.** A bfloat16 `KerasTensor` gets a float32 spec. A model that applies `ops.normalize` to a bfloat16 `Input` shows float32 in the graph, but it returns bfloat16 at run time. With an int32 `Input`, the graph shows float32 and the model returns int32 zeros.
4. **The new gradient test passes with an incorrect gradient.** See the comment on lines 2736 to 2783.

Item 3 is outside of the diff. Please make this change in the `Normalize` class:

```python
    def compute_output_spec(self, x):
        if self.axis is not None:
            canonicalize_axes(self.axis, len(x.shape))
        if backend.is_float_dtype(x.dtype):
            dtype = x.dtype
        else:
            dtype = backend.result_type(x.dtype, backend.floatx())
        return KerasTensor(shape=x.shape, dtype=dtype)
```

Keep the compute dtype and the output dtype separate. The float32 accumulation is correct, and `_rms_normalization` and `_layer_normalization` use the same method. The output dtype must stay equal to the input dtype. For bfloat16, the reason for the float32 accumulation is precision. bfloat16 has the exponent range of float32.

For an input that is not a float, use `backend.result_type(x.dtype, backend.floatx())` as the output dtype. This follows the configured `floatx`, and it applies the same 64-bit to 32-bit rule as the other ops. Do not use a literal `"float32"`, because it ignores the configuration. Do not use `backend.result_type(x.dtype, float)` either. That call takes the precision from the last two characters of `floatx()`, so a `bfloat16` configuration gives float16. float16 has the small exponent range that causes #23546.

Please also add a dtype test to `NNOpsDtypeTest` after `test_layer_normalization`. This test fails with the current code for the int32, bool, float16 and bfloat16 cases:

```python
    @parameterized.named_parameters(
        named_product(dtype=FLOAT_DTYPES + ["int32", "bool"], order=[1, 2])
    )
    def test_normalize(self, dtype, order):
        x = knp.ones((2, 8), dtype=dtype)
        if dtype in self.FLOAT_DTYPES:
            expected_dtype = dtype
        else:
            expected_dtype = backend.result_type(dtype, backend.floatx())

        self.assertDType(knn.normalize(x, order=order), expected_dtype)
        self.assertDType(
            knn.Normalize(order=order).symbolic_call(x), expected_dtype
        )
```

One more point, but not a blocker: a float16 zero vector still gets an `inf` gradient with the default `epsilon`. The clamped gradient is `1 / epsilon`, and this value is larger than the maximum float16 value. The code before this PR gives `NaN` for the same case on the torch and jax backends. This PR does not make the case worse.

The CPU test jobs for this PR are not complete. A new test run starts after your next push.

**Comment by indrajeetapache** [comment]
@JyotinderSingh Thanks for the detailed review. I’ve applied the suggested changes: separate compute/output dtypes for all orders, fixed compute_output_spec, strengthened the half-precision gradient test, and added the dtype test. Please take another look when you have a chance.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Thanks.

Looks like there is a conflict and you'll need to rebase on the latest master.

`keras/src/ops/nn.py` has a conflict in `_normalize`. Master moved the backend functions to `backend.ops`, so use `backend.ops.cast`, `backend.ops.linalg.norm`, `backend.ops.numpy.maximum` and `backend.ops.numpy.divide` in the new code. `backend.cast` no longer exists on master. The other helpers (`backend.is_float_dtype`, `backend.result_type`, `backend.floatx` and `backend.standardize_dtype`) are unchanged.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Another fix you might need to do based on some of the other code that has gone in: For an input that is not a float, use `backend.floatx()` as the output dtype, in `_normalize`, in `Normalize.compute_output_spec` and as the expected dtype in the new `NNOpsDtypeTest.test_normalize`. This is the rule in #23702, #23708 and #23545. `backend.result_type(x.dtype, backend.floatx())` gives float32 under `floatx="float64"` on the torch and numpy backends, where `floatx()` gives float64. I know I had asked for the `result_type` form in the first review, this replaces it.
